// react-automata/lib/State.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shouldHide = exports.shouldShow = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _minimatch = require('minimatch');

var _minimatch2 = _interopRequireDefault(_minimatch);

var _createConditional = require('./createConditional');

var _createConditional2 = _interopRequireDefault(_createConditional);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var displayName = 'State';

var contextTypes = {
  machineState: _propTypes2.default.string
};

var propTypes = {
  children: _propTypes2.default.node,
  value: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.string), _propTypes2.default.string]),
  onEnter: _propTypes2.default.func,
  onLeave: _propTypes2.default.func
};

var matches = function matches(value, machineState) {
  return machineState && (Array.isArray(value) ? value.some(function (state) {
    return (0, _minimatch2.default)(machineState, state);
  }) : (0, _minimatch2.default)(machineState, value));
};

var initial = function initial(props, context) {
  return matches(props.value, context.machineState);
};

var shouldShow = exports.shouldShow = function shouldShow(props, context) {
  return matches(props.value, context.machineState);
};

var shouldHide = exports.shouldHide = function shouldHide(props, context) {
  return !matches(props.value, context.machineState);
};

exports.default = (0, _createConditional2.default)({
  displayName: displayName,
  contextTypes: contextTypes,
  propTypes: propTypes,
  initial: initial,
  shouldShow: shouldShow,
  shouldHide: shouldHide
});
